File: editor/src/types/novel.ts
// Типы данных — полностью совпадают с Dart-моделями Amoria + формат v2
// (guides/format-v2.md, Часть 1). Все новые поля v2 — опциональные.

export interface EndingMetaEntry {
  id: string;
  title: string;
  hidden?: boolean;
}

export type StatIcon = 'heart' | 'star' | 'flame' | 'diamond' | 'moon' | 'sun' | 'leaf';

export interface StatDisplay {
  variable: string;
  label: string;
  icon?: StatIcon;
  color?: string; // hex "#E91E63"
  max?: number;
}

export interface PlayerNamePrompt {
  enabled: boolean;
  prompt?: string;
  defaultName?: string;
}

export interface NovelMeta {
  id: string;
  title: string;
  description: string;
  author: string;
  coverImage?: string;
  tags: string[];
  chaptersCount: number;
  sourceLanguage?: string; // язык оригинала (ru, en, etc.)
  dialogueTheme?: 'ornate' | 'artDeco' | 'modern' | 'glassmorphism' | 'fantasy'
    | 'victorian' | 'gothic' | 'noir' | 'sakura' | 'celestial'
    | 'cyberpunk' | 'steampunk' | 'pirate' | 'medieval' | 'egyptian'
    | 'baroque' | 'romantic' | 'nordic' | 'tropical' | 'bloodMoon';
  dialogueStyle?: 'classic' | 'center'; // classic=bottom, center=Romance Club style
  dialogueFrameColor?: string; // hex e.g. "#B8860B"
  dialogueBgColor?: string;    // hex e.g. "#1A1410"
  // v2 (1.3): список всех концовок для галереи «N из M»
  endings?: EndingMetaEntry[];
  // v2 (1.9): панель отношений
  statsDisplay?: StatDisplay[];
  // v2 (1.4): запрос имени игрока при первом старте
  playerNamePrompt?: PlayerNamePrompt;
  // v2.1 (4.1): версия формата (пишется при экспорте) и минимальная версия приложения
  formatVersion?: number;
  minAppVersion?: string;
}

// Текущая версия формата, проставляется в meta.json при экспорте (спека 4.1)
export const CURRENT_FORMAT_VERSION = 2;

export interface Character {
  id: string;
  name: string;
  color: string;
  sprites: CharacterSprite[];
  outfits?: Outfit[];
}

export interface CharacterSprite {
  id: string;
  image: string;
  label: string;
}

// v2 (1.5): гардероб. Формат строго по спеке:
// sprites — маппинг спрайт-ключ → путь картинки ("sprites/mia/casual_happy.png").
export interface Outfit {
  id: string;
  name: string;
  default?: boolean;
  priceDiamonds?: number;
  thumbnail?: string;
  sprites: Record<string, string>;
}

export interface Chapter {
  id: string;
  title: string;
  number: number;
  firstSceneId: string;
  scenes: Scene[];
  // v2 (1.8): рекап «Ранее…» перед первой сценой главы
  recap?: string;
}

// v2 (1.2): ветвление по переменным в конце сцены
export interface SceneBranch {
  conditions: Condition[];
  conditionsLogic?: ConditionsLogic;
  nextSceneId: string;
}

// v2 (1.3): концовка на сцене
export interface SceneEnding {
  id: string;
  title: string;
  description?: string;
  image?: string; // напр. "cg/ending_good.png"
}

export interface Scene {
  id: string;
  background?: string;
  music?: string;
  transition?: SceneTransition;
  backgroundLayers?: BackgroundLayer[];
  charactersOnScreen: SceneCharacter[];
  events: SceneEvent[];
  nextSceneId?: string;
  // v2 (1.2): ветки проверяются по порядку в конце сцены
  branches?: SceneBranch[];
  // v2 (1.3): достижение конца сцены = концовка
  ending?: SceneEnding;
  /** Позиция ноды в редакторе графа сцен. Сохраняется между переключениями глав
   *  и сессиями. Не экспортируется в рантайм-бандл новеллы (игнорируется клиентом). */
  editorPosition?: { x: number; y: number };
}

export interface SceneTransition {
  type: 'fade' | 'slideLeft' | 'slideRight' | 'dissolve' | 'none';
  duration: number; // мс
}

export interface BackgroundLayer {
  image: string;
  depth: number; // 0.0 (задний план) — 1.0 (передний)
  offsetX: number;
  offsetY: number;
}

export interface SceneCharacter {
  characterId: string;
  spriteId: string;
  position: 'left' | 'center' | 'right';
  animation?: 'fade_in' | 'fade_out' | 'slide_in_left' | 'slide_in_right' | 'bounce' | 'shake';
}

export type EventType = 'dialogue' | 'narration' | 'choice' | 'setVariable' | 'playSound' | 'changeBackground' | 'changeSprite' | 'effect' | 'showCg' | 'cameraMove' | 'showEmotion';

export type EffectType = 'shake' | 'flash' | 'fadeToBlack' | 'rain' | 'snow' | 'particles';
export type CgTransition = 'fade' | 'zoomIn';
export type EmotionType = 'heart' | 'sweatDrop' | 'question' | 'exclamation' | 'anger' | 'sparkle' | 'musicNote' | 'zzz';

export interface SceneEvent {
  type: EventType;
  speaker?: string;
  text?: string;
  choices?: Choice[];
  // setVariable
  variable?: string;
  value?: string | number | boolean;
  // Путь к ассету: имя файла фона для changeBackground ("forest.png"),
  // путь к звуку для playSound ("sounds/door.mp3"). Клиент читает это поле.
  asset?: string;
  characterId?: string;
  spriteId?: string;
  // Поля для события effect
  effectType?: EffectType;
  effectDuration?: number; // мс
  effectIntensity?: number; // 0.0–1.0
  // CG-арт
  cgImage?: string;
  cgTransition?: CgTransition;
  cgDuration?: number; // мс
  // Камера
  zoom?: number; // 0.5–2.0
  panX?: number;
  panY?: number;
  cameraDuration?: number; // мс
  // Эмоции
  emotionType?: EmotionType;
  // v2 (1.7): эмоция картинкой вместо emoji ("emotions/love.png")
  image?: string;
  // Cross-fade спрайтов
  spriteDuration?: number; // мс
  // Таймер на выбор
  timeLimit?: number; // секунды
  defaultChoiceIndex?: number;
  // v2 (1.6): озвучка dialogue/narration ("voice/ch1/mia_001.mp3")
  voice?: string;
}

export type ConditionsLogic = 'and' | 'or';

export interface Choice {
  text: string;
  nextSceneId: string;
  effects?: Record<string, string | number | boolean>;
  condition?: Condition;
  // v2 (1.1): составные условия; при заданных conditions приоритет у них
  conditions?: Condition[];
  conditionsLogic?: ConditionsLogic;
  // v2 (1.5): сюжетная разблокировка аутфитов ["mia:gala"]
  unlockOutfits?: string[];
  premium?: boolean;
  cost?: number;
}

export interface Condition {
  variable: string;
  operator: '>=' | '<=' | '==' | '!=' | '>' | '<';
  // v2 (1.1): в условиях допускаются и bool-значения ({ "value": true })
  value: number | boolean;
}

/// Перевод книги на один язык
export interface NovelTranslation {
  meta: {
    language: string;
    sourceLanguage: string;
    novelId: string;
    version: number;
  };
  novel?: { title?: string; description?: string };
  characters?: Record<string, { name?: string }>;
  chapters?: Record<string, { title?: string }>;
  texts: Record<string, string>; // оригинал → перевод
}

// Полное состояние проекта новеллы в редакторе
export interface NovelProject {
  meta: NovelMeta;
  characters: Character[];
  variables: Record<string, string | number | boolean>;
  chapters: Chapter[];
  translations?: Record<string, NovelTranslation>; // lang code → translation
}

File: client/lib/models/novel.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:equatable/equatable.dart';
import 'scene.dart';

part 'novel.g.dart';

/// Глава новеллы
@JsonSerializable()
class Chapter extends Equatable {
  final String id;
  final String title;
  final int number;
  final List<Scene> scenes;
  final String firstSceneId;

  /// v2: текст «Ранее…» — показывается один раз перед первой сценой главы
  final String? recap;

  const Chapter({
    required this.id,
    required this.title,
    required this.number,
    required this.scenes,
    required this.firstSceneId,
    this.recap,
  });

  factory Chapter.fromJson(Map<String, dynamic> json) =>
      _$ChapterFromJson(json);
  Map<String, dynamic> toJson() => _$ChapterToJson(this);

  Scene? getScene(String sceneId) {
    try {
      return scenes.firstWhere((s) => s.id == sceneId);
    } catch (_) {
      return null;
    }
  }

  @override
  List<Object?> get props => [id, title, number, scenes, firstSceneId, recap];
}

/// v2: описание концовки в meta.json (для галереи «N из M»)
@JsonSerializable()
class NovelEnding extends Equatable {
  final String id;
  final String title;
  final bool hidden;

  const NovelEnding({
    required this.id,
    this.title = '',
    this.hidden = false,
  });

  factory NovelEnding.fromJson(Map<String, dynamic> json) =>
      _$NovelEndingFromJson(json);
  Map<String, dynamic> toJson() => _$NovelEndingToJson(this);

  @override
  List<Object?> get props => [id, title, hidden];
}

/// v2: конфиг отображения стата в панели отношений (meta.statsDisplay)
@JsonSerializable()
class StatDisplayConfig extends Equatable {
  final String variable;
  final String label;

  /// heart | star | flame | diamond | moon | sun | leaf
  final String? icon;
  final String? color; // HEX
  final num max;

  const StatDisplayConfig({
    required this.variable,
    this.label = '',
    this.icon,
    this.color,
    this.max = 100,
  });

  factory StatDisplayConfig.fromJson(Map<String, dynamic> json) =>
      _$StatDisplayConfigFromJson(json);
  Map<String, dynamic> toJson() => _$StatDisplayConfigToJson(this);

  @override
  List<Object?> get props => [variable, label, icon, color, max];
}

/// v2: запрос имени игрока при первом старте новеллы
@JsonSerializable()
class PlayerNamePrompt extends Equatable {
  final bool enabled;
  final String? prompt;
  final String? defaultName;

  const PlayerNamePrompt({
    this.enabled = false,
    this.prompt,
    this.defaultName,
  });

  factory PlayerNamePrompt.fromJson(Map<String, dynamic> json) =>
      _$PlayerNamePromptFromJson(json);
  Map<String, dynamic> toJson() => _$PlayerNamePromptToJson(this);

  @override
  List<Object?> get props => [enabled, prompt, defaultName];
}

/// Метаданные новеллы
@JsonSerializable()
class NovelMeta extends Equatable {
  final String id;
  final String title;
  final String description;
  final String author;
  final String? coverImage;
  final String? coverUrl;
  final List<String> tags;
  // Back-compat: старые meta.json (и editor exporter <= v1) использовали "totalChapters".
  // Сервер и новый клиент пишут "chaptersCount". Принимаем оба ключа при чтении.
  @JsonKey(readValue: _readChaptersCount)
  final int chaptersCount;
  final int releasedChapters;
  final String? dialogueTheme;

  /// Цвет рамки диалога (hex, например "#B8860B"). Если задан — перекрывает тему.
  final String? dialogueFrameColor;

  /// Цвет фона диалога (hex, например "#1A1410"). Если задан — перекрывает тему.
  final String? dialogueBgColor;

  /// Стиль позиционирования диалога: "classic" (снизу), "center" (по центру как в Romance Club)
  final String? dialogueStyle;

  /// Локализованное название (заполняется из перевода если доступен)
  @JsonKey(includeFromJson: false, includeToJson: false)
  final String? localizedTitle;

  /// Локализованное описание (заполняется из перевода если доступен)
  @JsonKey(includeFromJson: false, includeToJson: false)
  final String? localizedDescription;

  /// v2: список всех концовок новеллы (для прогресса «N из M»)
  final List<NovelEnding> endings;

  /// v2: статы для панели отношений
  final List<StatDisplayConfig> statsDisplay;

  /// v2: запрос имени игрока при первом старте
  final PlayerNamePrompt? playerNamePrompt;

  const NovelMeta({
    required this.id,
    required this.title,
    required this.description,
    required this.author,
    this.coverImage,
    this.coverUrl,
    this.tags = const [],
    this.chaptersCount = 0,
    this.releasedChapters = 0,
    this.dialogueTheme,
    this.dialogueFrameColor,
    this.dialogueBgColor,
    this.dialogueStyle,
    this.localizedTitle,
    this.localizedDescription,
    this.endings = const [],
    this.statsDisplay = const [],
    this.playerNamePrompt,
  });

  factory NovelMeta.fromJson(Map<String, dynamic> json) =>
      _$NovelMetaFromJson(json);
  Map<String, dynamic> toJson() => _$NovelMetaToJson(this);

  /// Возвращает отображаемое название (локализованное если есть, иначе оригинал)
  String get displayTitle => localizedTitle ?? title;

  /// Возвращает отображаемое описание (локализованное если есть, иначе оригинал)
  String get displayDescription => localizedDescription ?? description;

  /// Тема диалогового окна
  DialogueFrameTheme get frameTheme {
    switch (dialogueTheme) {
      case 'artDeco': return DialogueFrameTheme.artDeco;
      case 'modern': return DialogueFrameTheme.modern;
      case 'glassmorphism': return DialogueFrameTheme.glassmorphism;
      case 'fantasy': return DialogueFrameTheme.fantasy;
      case 'victorian': return DialogueFrameTheme.victorian;
      case 'gothic': return DialogueFrameTheme.gothic;
      case 'noir': return DialogueFrameTheme.noir;
      case 'sakura': return DialogueFrameTheme.sakura;
      case 'celestial': return DialogueFrameTheme.celestial;
      case 'cyberpunk': return DialogueFrameTheme.cyberpunk;
      case 'steampunk': return DialogueFrameTheme.steampunk;
      case 'pirate': return DialogueFrameTheme.pirate;
      case 'medieval': return DialogueFrameTheme.medieval;
      case 'egyptian': return DialogueFrameTheme.egyptian;
      case 'baroque': return DialogueFrameTheme.baroque;
      case 'romantic': return DialogueFrameTheme.romantic;
      case 'nordic': return DialogueFrameTheme.nordic;
      case 'tropical': return DialogueFrameTheme.tropical;
      case 'bloodMoon': return DialogueFrameTheme.bloodMoon;
      default: return DialogueFrameTheme.ornate;
    }
  }

  /// Создать копию с переводом
  NovelMeta copyWithTranslation(String? translatedTitle, String? translatedDescription) {
    return NovelMeta(
      id: id,
      title: title,
      description: description,
      author: author,
      coverImage: coverImage,
      coverUrl: coverUrl,
      tags: tags,
      chaptersCount: chaptersCount,
      releasedChapters: releasedChapters,
      dialogueTheme: dialogueTheme,
      dialogueFrameColor: dialogueFrameColor,
      dialogueBgColor: dialogueBgColor,
      dialogueStyle: dialogueStyle,
      localizedTitle: translatedTitle,
      localizedDescription: translatedDescription,
      endings: endings,
      statsDisplay: statsDisplay,
      playerNamePrompt: playerNamePrompt,
    );
  }

  @override
  List<Object?> get props => [id, title, description, author, coverImage, coverUrl, tags, chaptersCount, releasedChapters, dialogueTheme, dialogueFrameColor, dialogueBgColor, dialogueStyle, localizedTitle, localizedDescription, endings, statsDisplay, playerNamePrompt];
}

/// Читает количество глав из JSON, поддерживая legacy-ключ "totalChapters".
Object? _readChaptersCount(Map map, String key) {
  return map[key] ?? map['totalChapters'];
}

/// Информация о главе с сервера
class ChapterInfo {
  final int number;
  final String title;
  final bool isReleased;
  final DateTime? releasedAt;
  bool isDownloaded;

  ChapterInfo({
    required this.number,
    required this.title,
    this.isReleased = false,
    this.releasedAt,
    this.isDownloaded = false,
  });

  factory ChapterInfo.fromJson(Map<String, dynamic> json) => ChapterInfo(
        number: json['number'] as int,
        title: (json['title'] as String?) ?? 'Глава ${json['number']}',
        isReleased: json['isReleased'] as bool? ?? false,
        releasedAt: json['releasedAt'] != null
            ? DateTime.parse(json['releasedAt'] as String)
            : null,
      );
}
